### PR TITLE
[4.0] Static exclusivity diagnostics handling of non-escaping closures passed as blocks.

### PR DESCRIPTION
**Pull request by atrick**
CCC Info:

Explanation: Static exclusivity diagnostics and its supporting analysis whitelist SIL patterns to ensure the diagnostic is sound. This needed to be expanded to handle the unusual case of nested non-escaping closures passed as blocks.

Scope of Issue: This broadens an assert that was recently added on the 4.0 branch. It avoids compiler crashes caused by that assert.

Radar: <rdar://problem/32928233> [Exclusivity] Fix static diagnostics to handle non-escaping '@convention(block)' arguments.

Risk: This change strictly decreases risk of compiler crashes. The risk is that code that breaks the exclusivity model could now pass diagnostics, but this possibility is almost inconceivable.

Testing: All usual tests on swift-4.0-branch. Benchmarks. Source compatibility. Other.

### CODE REVIEW

**Comment by atrick** [comment]
@swift-ci test.

**Comment by atrick** [comment]
@swift-ci Test Source Compatibility.

**Comment by atrick** [comment]
@devincoughlin I think you already reviewed these. Can you stamp this PR for CCC?

File: test/SILOptimizer/access_enforcement_noescape.swift [hunk 1]
```diff
@@ -504,9 +498,8 @@ func writeBoxWriteInout() {
 // CHECK-LABEL: } // end sil function '_T027access_enforcement_noescape18writeBoxWriteInoutyyFyycfU_'
 
 // Error: Cannot capture inout
-// Verification disabled because it suppresses other errors.
+// This triggers an early diagnostics, so it's handled in inout_capture_disgnostics.swift.
```
Comment (devincoughlin) [inline test/SILOptimizer/access_enforcement_noescape.swift]: Typo "an diagnostics". Same in comment above.

**Review by devincoughlin — APPROVED** [review]
Looks good to me. I noted a typo in a comment, but that doesn't need to be addressed for 4.0 CCC.

**Pull request merged**